### PR TITLE
Header support for C++ API

### PR DESCRIPTION
**Pull request by davidtrihy**
Producing and consuming header support in the C++ API.

Two test cases added in the e2e tests, was trying to add test cases for header support for empty/null messages but it appears that is not supported and the headers do not get attached. Is this correct behaviour?

Reference issue https://github.com/edenhill/librdkafka/issues/1861

### CODE REVIEW

File: examples/kafkatest_verifiable_client.cpp [hunk 1]
```diff
@@ -457,6 +457,17 @@ void msg_consume(RdKafka::KafkaConsumer *consumer,
               " [" << (int)msg->partition() << "]  at offset " <<
               msg->offset() << std::endl;
 
+        RdKafka::Headers *headers = msg->get_headers();
```
Comment (edenhill) [inline examples/kafkatest_verifiable_client.cpp]: There is no need for headers support in the verifiable_client, please remove these changes.

kafkatest_verifiable_client is used to run the Java system tests with the librdkafka client.

File: src-cpp/HeadersImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+#include <iostream>
+#include <string>
+#include <list>
+#include <cerrno>
+
+#include "rdkafkacpp_int.h"
+
+RdKafka::Headers  *RdKafka::Headers::create(size_t initial_count, bool free_rd_headers) {
```
Comment (edenhill) [inline src-cpp/HeadersImpl.cpp]: two spaces after type, should only be one.

File: src-cpp/HeadersImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    return new RdKafka::HeadersImpl(initial_count, free_rd_headers);
+}
+
+RdKafka::Headers  *RdKafka::Headers::create(const std::vector<Header> &headers, bool free_rd_headers) {
+    if (headers.size() > 0) {
+        return new RdKafka::HeadersImpl(headers, free_rd_headers);
+    } else {
+        return 0;
```
Comment (edenhill) [inline src-cpp/HeadersImpl.cpp]: I would've assumed an empty Headers list here rather than NULL

File: src-cpp/ProducerImpl.cpp [hunk 4]
```diff
@@ -143,14 +143,19 @@ RdKafka::ProducerImpl::produce (RdKafka::Topic *topic,
...
                                 const void *key, size_t key_len,
-                                int64_t timestamp,
-                                void *msg_opaque) {
+                                int64_t timestamp, void *msg_opaque,
+                                RdKafka::Headers *headers) {
+  rd_kafka_headers_t *hdrs;
+  if (headers) {
+    hdrs = headers->c_headers();
+  } else {
+    hdrs = 0;
```
Comment (edenhill) [inline src-cpp/ProducerImpl.cpp]: use NULL for pointers, not 0. And in this case you can initialize `hdrs` to NULL directly.

File: src-cpp/rdkafkacpp.h [hunk 5]
```diff
@@ -1365,7 +1366,139 @@ class RD_EXPORT MessageTimestamp {
...
+           RdKafka::ErrorCode err = ERR_NO_ERROR):
+    key_(key), err_(err) {
+        value_container_.assign(value);
+    };
+    
+    std::string key() const {
+        return key_;
+    }
+
+    const char* value() const {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: `const char *value..`

File: src-cpp/rdkafkacpp.h [hunk 6]
```diff
@@ -1365,7 +1366,139 @@ class RD_EXPORT MessageTimestamp {
...
+   * This object represents a single Header with key value pair
+   * and an ErrorCode
+   *
+   * @remark dynamic allocation of this object is not supported.
+   *
+   */
+  class Header {
+   public:
+    Header(const std::string& key,
+           const char* value,
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: `const char *value`

File: src-cpp/rdkafkacpp.h [hunk 7]
```diff
@@ -1365,7 +1366,139 @@ class RD_EXPORT MessageTimestamp {
...
+    key_(key), err_(err) {
+        value_container_.assign(value);
+    };
+    
+    std::string key() const {
+        return key_;
+    }
+
+    const char* value() const {
+        return value_container_.c_str();
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: values may be NULL

File: src-cpp/rdkafkacpp.h [hunk 8]
```diff
@@ -1365,7 +1366,139 @@ class RD_EXPORT MessageTimestamp {
...
+    }
+
+    RdKafka::ErrorCode err() const {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: What's the use of the error field?

File: src-cpp/rdkafkacpp.h [hunk 9]
```diff
@@ -1365,7 +1366,139 @@ class RD_EXPORT MessageTimestamp {
...
+    std::string key_;
+    RdKafka::ErrorCode err_;
+    std::string value_container_;
+    void *operator new(size_t); /* Prevent dynamic allocation */
+  };
+
+  /**
+   * @brief create a new instance of the Headers object
+   */
+  static Headers *create(size_t initial_size = 8, bool free_rd_headers = true);
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: Don't use default param values:
https://google.github.io/styleguide/cppguide.html#Default_Arguments

Document the  `@param`s.

I don't think free_rd_headers should be part of the public API

File: src-cpp/rdkafkacpp.h [hunk 10]
```diff
@@ -1365,7 +1366,139 @@ class RD_EXPORT MessageTimestamp {
...
+   *
+   * @remark Include <rdkafka/rdkafka.h> prior to including
+   *         <rdkafka/rdkafkacpp.h>
+   *
+   * @returns \c rd_kafka_headers_t*
+   */
+  virtual struct rd_kafka_headers_s *c_headers() = 0;
+
+  /** 
+   * @brief cleans up the underlying alocated C implementation headers if called
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: allocated

Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: when would you use this method?

File: src-cpp/rdkafkacpp.h [hunk 11]
```diff
@@ -1428,6 +1561,9 @@ class RD_EXPORT Message {
   /** @returns The \p msg_opaque as provided to RdKafka::Producer::produce() */
   virtual void               *msg_opaque () const = 0;
 
+  /** @returns The Headers instance for this Message (if applicable) */
+  virtual RdKafka::Headers   *get_headers() = 0;
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: Add new methods to end of the class.

File: src-cpp/rdkafkacpp.h [hunk 12]
```diff
@@ -2147,8 +2283,8 @@ class RD_EXPORT Producer : public virtual Handle {
                              int msgflags,
                              void *payload, size_t len,
                              const void *key, size_t key_len,
-                             int64_t timestamp,
-                             void *msg_opaque) = 0;
+                             int64_t timestamp, void *msg_opaque,
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: Must not change existing APIs since that will break the API stability guarantee

File: src-cpp/rdkafkacpp_int.h [hunk 13]
```diff
@@ -117,6 +116,122 @@ class EventImpl : public Event {
...
+    if (headers.size() > 0) {
+      headers_ = rd_kafka_headers_new(headers.size());
+      from_vector(headers);
+    } else {
+      headers_ = rd_kafka_headers_new(8);
+    }
+  }
+
+  ~HeadersImpl() {
+    if(free_headers_ && headers_) {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: `if (..`

File: src-cpp/rdkafkacpp_int.h [hunk 14]
```diff
@@ -117,6 +116,122 @@ class EventImpl : public Event {
...
+      rd_kafka_headers_destroy(headers_);
+    }
+  }
+
+  ErrorCode add(const std::string& key, const char* value) {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: `*` pointer goes on the variable, not the type

File: src-cpp/rdkafkacpp_int.h [hunk 15]
```diff
@@ -117,6 +116,122 @@ class EventImpl : public Event {
...
+    rd_kafka_resp_err_t err;
+    err = rd_kafka_header_add(headers_,
+                              key.c_str(), key.size(),
+                              value, strlen(value));
+    return static_cast<RdKafka::ErrorCode>(err);
+  }
+
+  ErrorCode remove(const std::string& key) {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: `&key`

File: src-cpp/rdkafkacpp_int.h (hunk 14 above)
Comment (edenhill) [reply src-cpp/rdkafkacpp_int.h]: value should be const void *, not char. and thus needs a `size_t value_size`

Comment (edenhill) [reply src-cpp/rdkafkacpp_int.h]: Can add a convenience method that takes a std::string as value

File: src-cpp/rdkafkacpp_int.h [hunk 16]
```diff
@@ -117,6 +116,122 @@ class EventImpl : public Event {
...
+    return static_cast<RdKafka::ErrorCode>(err);
+  }
+
+  std::vector<Headers::Header> get(const std::string &key) const {
+    std::vector<Headers::Header> headers;
+    const void *value;
+    size_t size;
+    rd_kafka_resp_err_t err;
+    for (size_t idx = 0;
+         !(err = rd_kafka_header_get(headers_, idx, key.c_str(), &value, &size)) ;\
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: remove backslash

File: src-cpp/rdkafkacpp_int.h [hunk 17]
```diff
@@ -117,6 +116,122 @@ class EventImpl : public Event {
...
+      }
+    }
+    return headers;
+  }
+
+  Headers::Header get_last(const std::string& key) const {
+    const void *value;
+    size_t size;
+    rd_kafka_resp_err_t err;
+    err = rd_kafka_header_get_last (headers_, key.c_str(), &value, &size);
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: `get_last(..`

File: src-cpp/rdkafkacpp_int.h [hunk 18]
```diff
@@ -117,6 +116,122 @@ class EventImpl : public Event {
...
+
+  size_t size() const {
+    return rd_kafka_header_cnt(headers_);
+  }
+
+  struct rd_kafka_headers_s* c_headers() {
+    return headers_;
+  }
+
+  ErrorCode destroy_headers() {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: I don't think there is much value in having an error return on this method

File: src-cpp/rdkafkacpp_int.h [hunk 19]
```diff
@@ -144,10 +263,11 @@ class MessageImpl : public Message {
 
   /* Create errored message */
   MessageImpl (RdKafka::Topic *topic, RdKafka::ErrorCode err):
-  topic_(topic), free_rkmessage_(false), key_(NULL) {
+  topic_(topic), free_rkmessage_(false), key_(NULL), headers_(0) {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: NULL, not 0

File: src-cpp/rdkafkacpp_int.h [hunk 20]
```diff
@@ -144,10 +263,11 @@ class MessageImpl : public Message {
...
     rkmessage_ = &rkmessage_err_;
     memset(&rkmessage_err_, 0, sizeof(rkmessage_err_));
     rkmessage_err_.err = static_cast<rd_kafka_resp_err_t>(err);
+
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: remove whitespace diff

File: src-cpp/rdkafkacpp_int.h [hunk 21]
```diff
@@ -213,8 +337,18 @@ class MessageImpl : public Message {
...
   rd_kafka_message_t rkmessage_err_;
   mutable std::string *key_; /* mutable because it's a cached value */
+  RdKafka::Headers *headers_;
 
 private:
+  RdKafka::Headers* get_headers_from_rkmessage(rd_kafka_message_t *rkmessage) {
+    rd_kafka_headers_t *hdrsp;
+    rd_kafka_resp_err_t err;
+
+    if (rkmessage->len > 0 && !(err = rd_kafka_message_headers(rkmessage, &hdrsp))) {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: you need to detach the headers to make sure they dont get freed with the rkmessage

File: tests/0085-headers.cpp [hunk 22]
```diff
@@ -0,0 +1,219 @@
...
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <iostream>
+#include "testcpp.h"
+
+class MyCbs : public RdKafka::OffsetCommitCb, public RdKafka::EventCb {
```
Comment (edenhill) [inline tests/0085-headers.cpp]: This class is not required for this test

File: tests/0085-headers.cpp [hunk 23]
```diff
@@ -0,0 +1,219 @@
...
+                 << actual_key << "' but got '" << expected_key << "'\n");
+    }
+    if (actual_value != expected_value) {
+      Test::Fail(tostr() << "Header value does not match, expected '" 
+                 << actual_value << "' but got '" << expected_value << "'\n");
+    }
+  }
+}
+
+static void test_n_headers (int n, const char* message) {
```
Comment (edenhill) [inline tests/0085-headers.cpp]: n -> num_hdrs or hdrcnt

File: tests/0085-headers.cpp [hunk 24]
```diff
@@ -0,0 +1,219 @@
...
+  std::string topic = Test::mk_topic_name("0085-headers", 1);
+  RdKafka::Conf *conf;
+  std::string errstr;
+
+  Test::conf_init(&conf, NULL, 0);
+
+  Test::conf_set(conf, "group.id", topic);
+  Test::conf_set(conf, "group.id", topic);
+  Test::conf_set(conf, "socket.timeout.ms", "10000");
```
Comment (edenhill) [inline tests/0085-headers.cpp]: remove all these, they're not needed for this test.

File: tests/0085-headers.cpp [hunk 25]
```diff
@@ -0,0 +1,219 @@
...
+
+static void test_n_headers (int n, const char* message) {
+  std::string topic = Test::mk_topic_name("0085-headers", 1);
+  RdKafka::Conf *conf;
+  std::string errstr;
+
+  Test::conf_init(&conf, NULL, 0);
+
+  Test::conf_set(conf, "group.id", topic);
+  Test::conf_set(conf, "group.id", topic);
```
Comment (edenhill) [inline tests/0085-headers.cpp]: duplicate

File: tests/0085-headers.cpp [hunk 26]
```diff
@@ -0,0 +1,219 @@
...
+    Test::Fail("Failed to set event callback: " + errstr);
+
+  RdKafka::Producer *p = RdKafka::Producer::create(conf, errstr);
+  if (!p)
+    Test::Fail("Failed to create Producer: " + errstr);
+
+  RdKafka::ErrorCode err;
+
+  std::vector<RdKafka::Headers::Header> headers_arr;
+  for (int i = 0; i < n; ++i) {
```
Comment (edenhill) [inline tests/0085-headers.cpp]: Isn't it better to create the Headers object prior to this loop and appending directly to it instead of first constructing a vector to pass

File: tests/0085-headers.cpp [hunk 27]
```diff
@@ -0,0 +1,219 @@
...
+  if (!c)
+    Test::Fail("Failed to create KafkaConsumer: " + errstr);
+  
+  std::vector<std::string> topics;
+  topics.push_back(topic);
+  if ((err = c->subscribe(topics)))
+    Test::Fail("subscribe failed: " + RdKafka::err2str(err));
+
+  int cnt = 0;
+  while (!cbs.seen_commit || !cbs.seen_stats) {
```
Comment (edenhill) [inline tests/0085-headers.cpp]: these are not related to this test. Instead wait for the same number of messages you produced above.

File: tests/0085-headers.cpp [hunk 28]
```diff
@@ -0,0 +1,219 @@
...
+  std::vector<RdKafka::Headers::Header> headers_arr;
+  for (int i = 0; i < n; ++i) {
+    std::stringstream key_s;
+    key_s << "header_" << i;
+    std::string key = key_s.str();
+    std::stringstream val_s;
+    val_s << "value_" << i;
+    std::string val = val_s.str();
+    headers_arr.push_back(RdKafka::Headers::Header(key, val.c_str()));
+  }
```
Comment (edenhill) [inline tests/0085-headers.cpp]: should also test identical keys (different values) and NULL values.

Comment (edenhill) [reply tests/0085-headers.cpp]: ..and empty headers list

File: tests/0085-headers.cpp [hunk 29]
```diff
@@ -0,0 +1,219 @@
...
+  RdKafka::Headers *produce_headers = RdKafka::Headers::create(headers_arr, false);
+
+  err = p->produce(topic, 0,
+                    RdKafka::Producer::RK_MSG_COPY,
+                    (void *)message, message ? strlen(message) : 0,
+                    (void *)"key", 3, 0, NULL, produce_headers);
+
```
Comment (edenhill) [inline tests/0085-headers.cpp]: produce a message without headers to make sure that works as well

File: tests/0085-headers.cpp [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+        err = c->commitAsync(msg);
+        if (err)
+          Test::Fail("commitAsync() failed: " + RdKafka::err2str(err));
+        rd_sleep(1); /* Sleep to simulate slow processing, making sure
+                      * that the offset commit callback op gets
+                      * inserted on the consume queue in front of
+                      * the messages. */
+      }
+      RdKafka::Headers *headers = msg->get_headers();
+      if (!headers) {
```
Comment (edenhill) [inline tests/0085-headers.cpp]: should try this on a message without headers and see that nothing breaks.

File: tests/0085-headers.cpp [hunk 31]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+static void test_ten_headers () {
+  Test::Say("Test ten headers in consumed message.\n");
+  std::string val = "valid";
+  test_n_headers(10, val.c_str());
+}
+
+static void test_one_header_null_msg () {
+  Test::Say("Test one header in consumed message with a null value message.\n");
```
Comment (edenhill) [inline tests/0085-headers.cpp]: I don't think this very valuable, the message value / key are not really related to the headers

File: tests/test.c [hunk 32]
```diff
@@ -283,6 +284,7 @@ struct test tests[] = {
         _TEST(0083_cb_event, 0, TEST_BRKVER(0,9,0,0)),
         _TEST(0084_destroy_flags_local, TEST_F_LOCAL),
         _TEST(0084_destroy_flags, 0),
+        _TEST(0085_headers, 0, TEST_BRKVER(0,11,0,0)),
```
Comment (edenhill) [inline tests/test.c]: Add to CMakeLists.txt too.

This goes for HeaderImpl.cpp too.

File: src-cpp/HeadersImpl.cpp [hunk 33]
```diff
@@ -0,0 +1,49 @@
...
+#include <list>
+#include <cerrno>
+
+#include "rdkafkacpp_int.h"
+
+RdKafka::Headers  *RdKafka::Headers::create(size_t initial_count, bool free_rd_headers) {
+    return new RdKafka::HeadersImpl(initial_count, free_rd_headers);
+}
+
+RdKafka::Headers  *RdKafka::Headers::create(const std::vector<Header> &headers, bool free_rd_headers) {
```
Comment (edenhill) [inline src-cpp/HeadersImpl.cpp]: free_rd_headers is an implementation detail, remove it from the public API.

File: src-cpp/rdkafkacpp.h (hunk 5 above)
Comment (edenhill) [reply src-cpp/rdkafkacpp.h]: header values may be NULL, and are to be considered binary, so use const void * and a size_t length instead

File: src-cpp/rdkafkacpp.h (hunk 12 above)
Comment (davidtrihy) [reply src-cpp/rdkafkacpp.h]: Noted thank you, I'll overload the method

File: src-cpp/rdkafkacpp.h (hunk 10 above)
Comment (davidtrihy) [reply src-cpp/rdkafkacpp.h]: It's probably an artifact of some experimentation I was doing, with your comment mentioning detaching the headers from the rkmessage I could completely encapsulate the Headers within in this instance and let the destructor manage the lifetime of the rd_headers

File: tests/0085-headers.cpp (hunk 31 above)
Comment (davidtrihy) [reply tests/0085-headers.cpp]: Just on this one when I produce messages with null or empty values the headers do not come through on the consume, is this expected behaviour?

File: src-cpp/rdkafkacpp.h (hunk 8 above)
Comment (davidtrihy) [reply src-cpp/rdkafkacpp.h]: The error field was just so when you got one individual Header when it was returned you could pass back if there was an error or not on the Header

File: src-cpp/rdkafkacpp.h (hunk 10 above)
Comment (davidtrihy) [reply src-cpp/rdkafkacpp.h]: Actually come to think of it, I think this is still needed in the case where producev fails and doesn't clean up the C headers and in that case you will have to manually clean them up yourself because producev can't set the internal C headers ptr to NULL so the destructor will fail for a double free

**Comment by alexander-alvarez** [comment]
@davidtrihy  thanks so much for pushing this forward! I know you're also looking to get this upstream in `node-rdkafka` will you have time to address the feedback and conflicts?

**Comment by alexander-alvarez** [comment]
bump @davidtrihy

**Comment by davidtrihy** [comment]
@alexander-alvarez sorry I've been busy with my day job around this and we're working with the forked version at the moment, I nearly have all the changes requested done but I'm just trying to get the tests working properly again. I'll try and get it done in the next week when I hopefully have some spare time.

**Comment by alexander-alvarez** [comment]
@davidtrihy friendly ping. I'd help out myself, but my C++ is only enough to read it and sort of understand what's going on.

**Comment by davidtrihy** [comment]
@alexander-alvarez Hey sorry been very busy but I'll spend some time at it tonight and tomorrow to address the comments, the node-rdkafka maintainer is happy with the changes on that side but I'll address the rest of the issues on this soon!

**Comment by alexander-alvarez** [comment]
@davidtrihy Friendly reminder from your neighborly open source C++ freeloader.

**Comment by davidtrihy** [comment]
@edenhill apologies about the lateness in getting the changes requested for this one, been very busy with work and I needed to find some time! I hope this suffices and if you have any further feedback let me know and I'll try and address ASAP so we can get this over the line.

**Comment by davidtrihy** [comment]
> @davidtrihy Friendly reminder from your neighborly open source C++ freeloader.

If there's any further feedback I'll address it ASAP so I can get this over the line

**Comment by alexander-alvarez** [comment]
Thanks @davidtrihy I owe you a :beer: or :cocktail: (or whatever you want lol) if you stop by NYC

**Comment by alexander-alvarez** [comment]
looks like there's more conflicts :(
@edenhill  can we get a look at this once they're fixed up?

File: src-cpp/rdkafkacpp.h [hunk 34]
```diff
@@ -1365,7 +1367,267 @@ class RD_EXPORT MessageTimestamp {
...
+   *         but if there is an error then it is the responsibility of the calling
+   *         object to deallocate the underlying C implementation if an instance
+   *         of the Headers object is created with free_rd_headers set to `false`
+   *
+   * @remark Include <rdkafka/rdkafka.h> prior to including
+   *         <rdkafka/rdkafkacpp.h>
+   *
+   * @returns \c rd_kafka_headers_t*
+   */
+  virtual struct rd_kafka_headers_s *c_headers() = 0;
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]:
```suggestion
  virtual struct rd_kafka_headers_s *c_ptrs() = 0;
```

File: src-cpp/rdkafkacpp.h [hunk 35]
```diff
@@ -1365,7 +1367,267 @@ class RD_EXPORT MessageTimestamp {
...
+
+  /** 
+   * @brief cleans up the underlying allocated C implementation headers if called
+   *
+   * @remark Safe to call even if the Headers object is set to clean up when
+   *         when the destructor is called
+   *
+   * @remark Safe to call even if the underlyng C pointer is set to null
+   */
+  virtual void destroy_headers() = 0;
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]:
```suggestion
  virtual void destroy_c_headers() = 0;
```

File: src-cpp/rdkafkacpp.h [hunk 36]
```diff
@@ -1451,6 +1713,9 @@ class RD_EXPORT Message {
    * @returns \c rd_kafka_message_t*
    */
   virtual struct rd_kafka_message_s *c_ptr () = 0;
+
+  /** @returns The Headers instance for this Message (if applicable) */
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: Beef this up. What does "if applicable" mean? What is returned if they are not applicable?

File: src-cpp/rdkafkacpp.h [hunk 37]
```diff
@@ -2147,8 +2412,18 @@ class RD_EXPORT Producer : public virtual Handle {
                              int msgflags,
                              void *payload, size_t len,
                              const void *key, size_t key_len,
-                             int64_t timestamp,
-                             void *msg_opaque) = 0;
+                             int64_t timestamp, void *msg_opaque) = 0;
+
+  /**
+   * @brief produce() variant that that allows for Header support on produce
+   *        Otherwise identical to produce() above.
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: need to explain ownership of Headers if produce succeeds or fails

File: src-cpp/rdkafkacpp_int.h [hunk 38]
```diff
@@ -117,25 +116,148 @@ class EventImpl : public Event {
...
+  }
+
+  struct rd_kafka_headers_s* c_headers() {
+    return headers_;
+  }
+
+  void destroy_headers() {
+    if (headers_) {
+      rd_kafka_headers_destroy(headers_);
+      headers_ = 0;
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]:
```suggestion
      headers_ = NULL;
```

File: src-cpp/rdkafkacpp_int.h [hunk 39]
```diff
@@ -117,25 +116,148 @@ class EventImpl : public Event {
...
+      }
+    }
+    return headers;
+  }
+
+  size_t size() const {
+    return rd_kafka_header_cnt(headers_);
+  }
+
+  struct rd_kafka_headers_s* c_headers() {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]:
```suggestion
  struct rd_kafka_headers_s  *c_headers() {
```

File: src-cpp/rdkafkacpp_int.h (hunk 38 above)
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]:
```suggestion
      headers_ = NULL;
```

File: src-cpp/rdkafkacpp.h [hunk 40]
```diff
@@ -1365,7 +1367,267 @@ class RD_EXPORT MessageTimestamp {
...
+     * @remark The error code is used for when the Header is constructed internally
+     *         by using something like RdKafka::Headers::get_last which constructs
+     *         a Header encapsulating the ErrorCode in the process
+     */
+    Header(const std::string &key,
+           const void *value,
+           size_t value_size,
+           const RdKafka::ErrorCode &err):
+    key_(key), err_(err), value_size_(value_size)  {
+        value_ = copy_value(value, value_size);
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]: It would be more efficient to just reference the underlying C header value data, if possible

File: src-cpp/rdkafkacpp_int.h [hunk 41]
```diff
@@ -206,15 +328,29 @@ class MessageImpl : public Message {
...
   rd_kafka_message_t rkmessage_err_;
   mutable std::string *key_; /* mutable because it's a cached value */
+  RdKafka::Headers *headers_;
 
 private:
+  RdKafka::Headers* get_headers_from_rkmessage(rd_kafka_message_t *rkmessage) {
+    rd_kafka_headers_t *hdrsp;
+    rd_kafka_resp_err_t err;
+
+    if (rkmessage->len > 0 && !(err = rd_kafka_message_detach_headers(rkmessage, &hdrsp))) {
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: You can remove the err variable, it is not used

File: src-cpp/rdkafkacpp_int.h [hunk 42]
```diff
@@ -206,15 +328,29 @@ class MessageImpl : public Message {
...
+  Headers* get_headers() {
+    return headers_;
+  }
+
   RdKafka::Topic *topic_;
   rd_kafka_message_t *rkmessage_;
   bool free_rkmessage_;
   /* For error signalling by the C++ layer the .._err_ message is
    * used as a place holder and rkmessage_ is set to point to it. */
   rd_kafka_message_t rkmessage_err_;
```
Comment (edenhill) [inline src-cpp/rdkafkacpp_int.h]: Was this meant to be a pointer? Having the entire struct here for an error code seems wasteful.

File: tests/0085-headers.cpp [hunk 43]
```diff
@@ -0,0 +1,395 @@
+/*
+ * librdkafka - Apache Kafka C library
+ *
+ * Copyright (c) 2012-2015, Magnus Edenhill
```
Comment (edenhill) [inline tests/0085-headers.cpp]:
```suggestion
 * Copyright (c) 2018, Magnus Edenhill
```

File: src-cpp/rdkafkacpp.h [hunk 44]
```diff
@@ -1451,6 +1713,9 @@ class RD_EXPORT Message {
    * @returns \c rd_kafka_message_t*
    */
   virtual struct rd_kafka_message_s *c_ptr () = 0;
+
+  /** @returns The Headers instance for this Message (if applicable) */
+  virtual RdKafka::Headers   *get_headers() = 0;
```
Comment (edenhill) [inline src-cpp/rdkafkacpp.h]:
```suggestion
  virtual RdKafka::Headers   *headers() = 0;
```

**Comment by alexander-alvarez** [comment]
@davidtrihy It's close to the finish line

**Comment by alexander-alvarez** [comment]
Bump @davidtrihy

**Comment by Fr8Traindb** [comment]
I am patiently waiting for this as well.  Any idea on when this will be completed?

**Comment by edenhill** [comment]
I'm picking this up to do the finishing touches

**Comment by edenhill** [comment]
Made the final changes and created a new PR here: https://github.com/edenhill/librdkafka/pull/2109

Please review and test out.
